packages/build-utils/src/transforms/remove-fenced-code.ts

```diff
@@ -0,0 +1,482 @@
+import { hasProperty } from '@metamask/utils';
+
+/**
+ * Two sets of feature labels, where:
+ * - `active` is the set of labels that are active for the current build.
+ * - `all` is the set of all labels that are declared in the codebase.
+ *
+ * For `ONLY_INCLUDE_IF` fences, the code fence removal transform will
+ * include the fenced code if any of the specified labels are active. See
+ * {@link removeFencedCode} for details.
+ */
+export type FeatureLabels = {
+  active: ReadonlySet<string>;
+  all: ReadonlySet<string>;
+};
+
+enum DirectiveTerminus {
+  BEGIN = 'BEGIN',
+  END = 'END',
+}
+
+export enum DirectiveCommand {
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  ONLY_INCLUDE_IF = 'ONLY_INCLUDE_IF',
+}
+
+// Matches lines starting with "///:", and any preceding whitespace, except
+// newlines. We except newlines to avoid eating blank lines preceding a fenced
+// line.
+// Double-negative RegEx credit: https://stackoverflow.com/a/3469155
+const linesWithFenceRegex = /^[^\S\r\n]*\/\/\/:.*$/gmu;
+
+// Matches the first "///:" in a string, and any preceding whitespace
+const fenceSentinelRegex = /^\s*\/\/\/:/u;
+
+// Breaks a fence directive into its constituent components.
+// At this stage of parsing, we are looking for one of:
+// - TERMINUS:COMMAND(PARAMS)
+// - TERMINUS:COMMAND
+const directiveParsingRegex =
+  /^([A-Z]+):([A-Z_]+)(?:\(((?:\w[-\w]*,)*\w[-\w]*)\))?$/u;
+
+/**
+ * Removes fenced code from the given JavaScript source string. "Fenced code"
+ * includes the entire fence lines, including their trailing newlines, and the
+ * lines that they surround.
+ *
+ * A valid fence consists of two well-formed fence lines, separated by one or
+ * more lines that should be excluded. The first line must contain a `BEGIN`
+ * directive, and the second most contain an `END` directive. Both directives
+ * must specify the same command.
+ *
+ * Here's an example of a valid fence:
+ *
+ * ```javascript
+ *   ///: BEGIN:ONLY_INCLUDE_IF(build-flask)
+ *   console.log('I am Flask.');
+ *   ///: END:ONLY_INCLUDE_IF
+ * ```
+ *
+ * For details, please see the documentation.
+ *
+ * @param filePath - The path to the file being transformed.
+ * @param fileContent - The contents of the file being transformed.
+ * @param featureLabels - FeatureLabels that are currently active.
+ * @returns A tuple of the post-transform file contents and a boolean indicating
+ * whether they were modified.
+ */
+export function removeFencedCode(
```

Comment:
Wondering if we should include the stream functionality in this package too or do you think the integration should live fully in the client repos?

Reply:
Added context in PR description:

> - The decision to only extract the `removeFencedCode` and `lintTransformedFile` functions was made for two reasons primarily:
>  - The integrations of code fence removal in [the extension / Browserify](https://github.com/MetaMask/metamask-extension/blob/develop/development/build/transforms/remove-fenced-code.js/#L12-L66) and [mobile / Metro](https://github.com/MetaMask/metamask-mobile/blob/3ba70c4d083ac01d63cdfc06c4257f47fa808cc2/metro.transform.js) are significantly different.
>  - The different applications have different ESLint configurations. It seems inappropriate to make this package responsible for being aware of the ESLint configs of the consumers.